x/valset-pref/validator_set.go

```diff
@@ -29,6 +29,99 @@ func (k Keeper) SetupValidatorSetPreference(ctx sdk.Context, delegator string, p
+	existingSet, found := k.GetValidatorSetPreference(ctx, delegatorAddr)
+	if !found {
+		return fmt.Errorf("user %s doesn't have validator set", delegatorAddr)
+	}
+
+	delegator, err := sdk.AccAddressFromBech32(delegatorAddr)
+	if err != nil {
+		return err
+	}
+
+	// loop through the validatorSetPreference and delegate the proportion of the tokens based on weights
+	for _, val := range existingSet.Preferences {
+		_, validator, err := k.getValAddrAndVal(ctx, val.ValOperAddress)
+		if err != nil {
+			return err
+		}
+
+		// tokenAmt takes the amount to delegate, calculated by {val_distribution_weight * tokenAmt}
+		// expect tokenAmt to be in the smallest denom for ex, 1 usmo = 10^6 osmo
+		tokenAmt := val.Weight.Mul(coin.Amount.ToDec()).RoundInt()
+
+		// TODO: What happens here if validator is jailed, tombstoned, or unbonding
+		// Delegate the unbonded tokens
+		_, err = k.stakingKeeper.Delegate(ctx, delegator, tokenAmt, stakingtypes.Unbonded, validator, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UndelegateFromValidatorSet undelegates {coin} amount from the validator set.
+// For ex: userA has staked 10tokens with weight {Val->0.5, ValB->0.3, ValC->0.2}
+// undelegate 6osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
+// our undelegate logic would attempt to undelegate 3osmo from A , 1.8osmo from B, 1.2osmo from C
+func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string, coin sdk.Coin) error {
+	// get the existing validator set preference
+	existingSet, found := k.GetValidatorSetPreference(ctx, delegatorAddr)
+	if !found {
+		return fmt.Errorf("user %s doesn't have validator set", delegatorAddr)
+	}
+
+	delegator, err := sdk.AccAddressFromBech32(delegatorAddr)
+	if err != nil {
+		return err
+	}
+
+	// the total amount the user wants to undelegate
+	tokenAmt := sdk.NewDec(coin.Amount.Int64())
+
+	totalAmountFromWeights := sdk.NewDec(0)
+	for _, val := range existingSet.Preferences {
+		// Calculate the amount to undelegate based on the existing weights
+		amountToUnDelegate := val.Weight.Mul(tokenAmt)
+
+		// ValidateValidator gurantees that this exist
+		valAddr, err := sdk.ValAddressFromBech32(val.ValOperAddress)
+		if err != nil {
+			return err
+		}
+
+		validator, found := k.stakingKeeper.GetValidator(ctx, valAddr)
+		if !found {
+			return fmt.Errorf("validator %s not found", valAddr)
+		}
+
+		sharesAmt, err := validator.SharesFromTokens(amountToUnDelegate.RoundInt())
+		if err != nil {
+			return err
+		}
+
+		_, err = k.stakingKeeper.Undelegate(ctx, delegator, valAddr, sharesAmt) // this has to be shares amount
+		if err != nil {
+			return err
+		}
+
+		totalAmountFromWeights = totalAmountFromWeights.Add(amountToUnDelegate)
+	}
+
+	if !totalAmountFromWeights.Equal(tokenAmt) {
+		return fmt.Errorf("The undelegate total donot add up with the amount calculated from weights expected %s got %s", tokenAmt, totalAmountFromWeights)
+	}
```

Comment:
To be clear, this error check is nothing but informative right? Bc regardless if it matches or not, the undelegations will have taken place

Reply:
ah yes! i'll put this logic before `undelegate`, the only downside would be to run the loop twice